src/main/java/net/neoforged/neoforge/event/BlockEntityTypesValidBlocksEvent.java

```diff
@@ -0,0 +1,112 @@
+import net.minecraft.world.level.block.WallSignBlock;
+import net.minecraft.world.level.block.entity.BlockEntityType;
+import net.neoforged.bus.api.Event;
+import net.neoforged.bus.api.SubscribeEvent;
+import net.neoforged.fml.ModLoader;
+import net.neoforged.fml.common.EventBusSubscriber;
+import net.neoforged.fml.event.IModBusEvent;
+import net.neoforged.fml.event.lifecycle.FMLCommonSetupEvent;
+import org.jetbrains.annotations.Nullable;
+
+public class BlockEntityTypesValidBlocksEvent extends Event implements IModBusEvent {
+    private final ResourceKey<BlockEntityType<?>> blockEntityTypeResourceKey;
+    private final BlockEntityType<?> blockEntityType;
+    private final Set<Block> currentValidBlocks;
+    @Nullable
+    private Class<?> baseClass = null;
+
+    protected BlockEntityTypesValidBlocksEvent(ResourceKey<BlockEntityType<?>> blockEntityTypeResourceKey, BlockEntityType<?> blockEntityType) {
+        this.blockEntityTypeResourceKey = blockEntityTypeResourceKey;
+        this.blockEntityType = blockEntityType;
+        this.currentValidBlocks = new HashSet<>(blockEntityType.getValidBlocks());
+    }
+
+    /**
+     * The {@link ResourceKey} that this current {@link BlockEntityType} was registered under.
+     */
+    public ResourceKey<BlockEntityType<?>> getBlockEntityTypeResourceKey() {
+        return blockEntityTypeResourceKey;
+    }
+
+    /**
+     * The current {@link BlockEntityType}.
+     */
+    public BlockEntityType<?> getBlockEntityType() {
+        return blockEntityType;
+    }
+
+    /**
+     * An immutable list of the list of valid blocks that will be assigned to replace {@link BlockEntityType#validBlocks}'s value.
+     */
+    public Set<Block> getCurrentValidBlocks() {
+        return Collections.unmodifiableSet(this.currentValidBlocks);
+    }
+
+    /**
+     * Will add the given block to the list of valid blocks we will assign to the current {@link BlockEntityType}.
+     * Make sure the given block is the same or derived from the highest common class of all the entries in valid blocks.
+     * <p></p>
+     * Example: If the valid blocks list has {@link StandingSignBlock} entry and {@link WallSignBlock} entry, the common class is SignBlock
+     * and the given block must be a {@link SignBlock} or has {@link SignBlock} as a child class in its hierarchy.
+     * 
+     * @param block The block to add as a valid block for the current {@link BlockEntityType}
+     */
+    public void addValidBlock(Block block) {
+        setCommonSuperClassForExistingValidBlocks();
+
+        if (this.baseClass == null || this.baseClass.isAssignableFrom(block.getClass())) {
+            this.currentValidBlocks.add(block);
+        } else {
+            throw new IllegalArgumentException("Given block " + block + " does not derive from existing valid block's common superclass of " + this.baseClass);
+        }
+    }
+
+    private void setCommonSuperClassForExistingValidBlocks() {
+        if (this.baseClass == null) {
```

Comment:
Given that the `baseClass` is only ever computed once, with the initial set of blocks, thanks to the `== null` check -- perhaps this should go in a final field on the event or at the very least a memoized supplier?